### PR TITLE
fix: React DOM crashes when <option> contains three interpolated valu…

### LINKED ISSUE #11911: React DOM crashes when <option> contains three interpolated value if one is a conditional.
**Do you want to request a *feature* or report a *bug*?** 

Bug 

**What is the current behavior?**

React DOM crashes when `<option>` contains three interpolated value if one is a conditional. 

Reproduction: https://jsfiddle.net/0opjvycp/ 

1. Change the value of the `<select>`
2. React crashes with `NotFoundError: Node was not found` 

From what I can tell, the conditional value is necessary, and it must be three interpolated values. 

**What is the expected behavior?**

React should not crash. 

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**

React DOM 16.2 and 16.0. I think this worked in 15.6 - https://jsfiddle.net/mrwkmuqc/ does not crash

### PR DESCRIPTION
**Pull request by stephenkingsley**
fix #11911 
…e if one is a conditional. #11911

**Before submitting a pull request,** please make sure the following is done:
1. [x] Fork [the repository](https://github.com/facebook/react) and create your branch from `master`.
2. [x] Run `yarn` in the repository root.
3. [x] If you've fixed a bug or added code that should be tested, add tests!
4. [x] Ensure the test suite passes (`yarn test`). Tip: `yarn test --watch TestName` is helpful in development.
5. [x] Run `yarn test-prod` to test in the production environment. It supports the same options as `yarn test`.
6. [x] If you need a debugger, run `yarn debug-test --watch TestName`, open `chrome://inspect`, and press "Inspect".
7. [x] Format your code with [prettier](https://github.com/prettier/prettier) (`yarn prettier`).
8. [x] Make sure your code lints (`yarn lint`). Tip: `yarn linc` to only check changed files.
9. [x] Run the [Flow](https://flowtype.org/) typechecks (`yarn flow`).
10. [x] If you haven't already, complete the CLA.

**Learn more about contributing:** https://reactjs.org/docs/how-to-contribute.html

### CODE REVIEW

File: packages/react-dom/src/__tests__/ReactDOMOption-test.js [hunk 1]
```diff
@@ -46,7 +46,7 @@ describe('ReactDOMOption', () => {
         '    in div (at **)\n' +
         '    in option (at **)',
     );
-    expect(node.innerHTML).toBe('1  2');
+    expect(node.innerHTML).toBe('1 <div></div> 2');
```
Comment (gaearon) [inline packages/react-dom/src/__tests__/ReactDOMOption-test.js]: This is a behavior change. I don't think we want this.

**Comment by stephenkingsley** [comment]
Sorry that is my fault! I will fix it in today.

**Comment by stephenkingsley** [comment]
@gaearon please review again

**Comment by stephenkingsley** [comment]
@gaearon Would you have anytime to review, please?

**Comment by gaearon** [comment]
Can you explain what this is doing? It’s not obvious from the code that it is correct. Generally when submitting a PR it is a good idea to explain how you arrived at this conclusion, when it got broken, whether you can point to React 15 code doing a similar thing, etc.

**Comment by stephenkingsley** [comment]
My english is not good enough to explain how I do this.

When render `<option>{1}<div />{2}</option>`, Option has three children like [textNode, Div, textNode], so I use `reduceOptionChild` function to remove Div. Because the Div is illegal in option. 

If render like this, when change the value, option has one child that is textNode. So react was crash and warning Node was not found. I find out function [`assertValidProps`](https://github.com/stephenkingsley/react/blob/2e542f3d177741add4030e393ddaa566f43dae60/packages/react-dom/src/client/ReactDOMFiberComponent.js#L885) change this option's child. Before change, this option has three childNodes. When run assertValidProps, the option has one childNode: textNode. So I use `span` contain textNodes.
```
<option key={value} value={value}>
  {isSelected && '(Y) '}
  {value} 
  {'Some string'}
</option>
```

I think you're right, I should do something like React 15 code.

**Comment by gaearon** [comment]
My point is:

* If this is a regression from React 15 behavior, please implement whatever React 15 did.
* If this has always been present then let's discuss possible solutions separately.

**Comment by aweary** [comment]
I only looked at this briefly, but I think there may be a deeper issue with child reconciliation. Specifically [this logic (or maybe somewhere else upstream) in `commitPlacement`](https://github.com/facebook/react/blob/master/packages/react-reconciler/src/ReactFiberCommitWork.js#L467-L479). `before` should always be a child of `parent`, and in this case it isn't. I'm not exactly sure why that's the case yet, but my gut feeling is that this solution is fixing the symptoms not the problem.

**Comment by gaearon** [comment]
This was fixed in https://github.com/facebook/react/pull/13261. Thanks!

**Pull request closed without merging**